### PR TITLE
[Woo POS] [Cash & Receipts] Disallow sending receipt and show toast when WC version is not eligible

### LINKED ISSUE #14465: [Woo POS] [Cash & Receipts] Prompt alert when Woo/WCPay versions are no correct
> I personally wouldn’t change eligibility criteria for the POS entry point for this project. I would show the button and show a snackbar/toast when the user clicks it “Please update Woo/WCPay to version xyz”.

### PR DESCRIPTION
**Pull request by iamgabrielma**
Closes: #14465

## Description
This PR adds eligibility checks to receipts usage, prompting the merchant to upgrade to WooCommerce 9.5 to generate receipts in POS.

We add a toast/notification to the "success payment" view displaying that message, which they can tap to dismiss.

https://github.com/user-attachments/assets/5d467730-aa1a-4829-85f4-663439cb79ed

## Testing information
- Set feature flag `.sendReceiptsForPointOfSale` to true
- Log into Povilas testing site `https://lovely-aware-sturgeon.jurassic.ninja/` details: PdfdoF-5Qj-p2
- Proceed with a POS transaction
- Observe that the modal to enter email for receipts works normally
- Switch to a non eligible site (Woo <9.5) or edit `minimumVersionForPOSReceipts = 10`
- Proceed with a POS transaction
- Observe that upon tapping `Receipt` the toast shows up 
- Tapping the toast should dismiss it.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:85]: I've just added the feature flag here to save a request to the system status store if not needed, but we can remove it if you feel is unnecessary, since the whole button is already under a feature flag 👍

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift]: I needed to remove this as it was causing a bug with the text rendering when adding new elements to the view, note how the text becomes extremely small when the view changes:


https://github.com/user-attachments/assets/46895d7b-edee-43ad-a6c6-63dd8a348505

After quite a bit I still haven't found the reason for this, but I cannot see either why we are allowing it to be scaled down to 10% of the original size, so I've removed it. Happy to iterate as needed.

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift]: I think let's move it to a separate file for better organization.

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift]: The whole view moves up for this toast. Usually, toasts appear on the top of the view. What do you think of putting it in ZStack instead?

Also, only the receipt button title animates while other views just snap. If we keep a solution without Z-stack, then it would be nice to fix this. 

https://github.com/user-attachments/assets/42a385ba-f678-4e4c-a574-b7ed5c338984

File: WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:381]: nit;

We could have something like `ReceiptEligibilityUseCase` for POS as well, so this code would be used but not kept within the aggregate model.

Comment (staskus) [reply WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:381]: Also, I'm thinking about whether _any_ of this code should be within the aggregate model at all. I think we haven't established strict rules and practices yet but it doesn't look like anything receipt-related is required to be used beyond the final screen.

We could have something like `PointOfSaleReceiptModel` which we could use within `TotalsView` or other receipt-related views. 

What do you think? I also wonder what's the view of @joshheald 🤔

File: WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
Comment (staskus) [reply WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:85]: Yes, I think it's fine both ways. We can keep it as it is, since the flag will be removed either way.

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: I did this together with other changes to make sure that the card reader and total views are prioritized over descriptions. It was more important that totals are shown. However, it was meant to kick in for extra-large accessibility sizes. Not for normal sizes.

I don't find anything too bad happening after removal.  For example, the buttons now don't fully fit after an extra large accessibility size is used. 

image

We can handle this in similar ways as we do with other views, and just hide an image for accessibility sizes (Like we do in `POSErrorXMark`). Such an approach makes sure that the most important information and CTAs fit:

image

Another approach could be to keep a similar behavior but only let it kick in if dynamic type is increased, for example:

```
@ScaledMetric private var scale: CGFloat = 1.0
<...>
    .minimumScaleFactor(1 / scale)
```

**Review by staskus — COMMENTED** [review]
Thanks for the work!

I left a couple of comments. Some of them are related to architecture, so I don't want them to become a blocker. It's interesting to me how we want to approach development of such functionality and I think in the future it will be clearer for all of us.

----

On a product note, I wonder if we even need such a functionality. Maybe using an `info` icon that we have on product list with an explanation of what features are missing due to outdated WooCommerce version rather than sprinkling non-functional CTAs 🤔 Not a blocker, just something to think about.

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: 👍  On https://github.com/woocommerce/woocommerce-ios/pull/14587/commits/20a9725b9b0834e23e68f7a428c6b15b3477a5fc I've modified this to follow `POSErrorXMark` approach, so it doesn't render on large sizes:

Simulator Screenshot - iPad mini (A17 Pro) - 2024-12-05 at 10 09 19

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: Done! https://github.com/woocommerce/woocommerce-ios/pull/14587/commits/785db31140a4ff38fb5d19348878a6d59093e35c

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: I had to play with this one a bit, as the ZStack was making the notice to not show on landscape, but do on portrait. Fixed it on https://github.com/woocommerce/woocommerce-ios/pull/14587/commits/b9f96bc0fe2dba5b33f2f3ffc3fe3dad897c8683

Simulator Screenshot - iPad mini (A17 Pro) - 2024-12-05 at 09 58 15

It shows a bit hidden for large accessibility sizes due to the floating connection button, but I think this is fine for now:

Simulator Screenshot - iPad mini (A17 Pro) - 2024-12-05 at 10 09 11

Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: > Also, only the receipt button title animates while other views just snap. If we keep a solution without Z-stack, then it would be nice to fix this.

Good eye! Do you think this is some sort of implicit animation coming from the parent? I could only address this by being explicit in the Text component about disabling it, fixed on https://github.com/woocommerce/woocommerce-ios/pull/14587/commits/1f7cc976764ab490a9dfd7920717df7ee40cd1ac

File: WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:381]: Yes, I'm not quite sure where this should be placed. If has to go through the aggregate model, perhaps would make sense to have all receipts-related logic on its own controller? If not, perhaps some service/util that we declare directly in the totals view?

At the moment only need to perform the check by sending an action to the dispatcher, so I don't think we need to make this eligibility check part of the aggregate model and could be a simple util declared in the same totals view, but I'm a bit on the fence depending on coming design updates, and to not pack views with business logic.

I've opened a follow-up here just to not block this PR: https://github.com/woocommerce/woocommerce-ios/issues/14609

**Comment by iamgabrielma** [comment]
Thanks for the review @staskus , addressed all items above! 👍

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [reply WooCommerce/Classes/POS/Presentation/TotalsView.swift]: > Good eye! Do you think this is some sort of implicit animation coming from the parent? 

Likely, but the Receipt button can continue animating, just together as one element rather than outline and title separately. I don't know fix helped and I can still see issues.

---------

However, the whole view continues to move with the toast, do you reproduce it? It shouldn't happen with `ZStack`. 

Also, wrap `isShowingReceiptsNotEligibleBanner` in `withAnimation` block or add `.animation(.default, value: isShowingReceiptsNotEligibleBanner)` to have a proper animation when this value changes.

https://github.com/user-attachments/assets/fccc48ac-375a-4100-8956-978b2f3df29d

File: WooCommerce/Classes/POS/Presentation/POSReceiptEligibilityBanner.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/POSReceiptEligibilityBanner.swift:22]: As mentioned, animate the appearance the same way

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift]: Lets leave it like in this in this scope of this PR, likely it will make sense to move it to success view p1733389827659689/1733375209.052419-slack-C070SJRA8DP.

We can see that all this logic changes when the success view logic changes.

File: WooCommerce/Classes/POS/Presentation/TotalsView.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/TotalsView.swift]: It doesn't solve animation issues for me, we can remove this and address animations later if you prefer.

**Review by staskus — APPROVED** [review]
Let's address success view jumping when the toast appears before merging.

I assume all of this code is likely to change due to our discussions of moving this logic to the success view.

**Comment by iamgabrielma** [comment]
> Let's address success view jumping when the toast appears before merging. I assume all of this code is likely to change due to our discussions of moving this logic to the success view.

👋  Circling back to this one after making the changes to the payment buttons on https://github.com/woocommerce/woocommerce-ios/pull/14622 , the only remaining issue here was to resolve the issues with animation, which don't happen anymore since we're using matching geometry for the buttons, and we've wrapped the banner in a Zstack so doesn't move the rest of components:

https://github.com/user-attachments/assets/0d9a982d-d83c-4714-801d-cfcb91bb694c

I'll leave this here for a bit before merging, also cc: @jaclync for our conversation regarding if receipt-logic should be in the aggregate model, it is here for now but I'm moving it out with https://github.com/woocommerce/woocommerce-ios/issues/14609

**Pull request merged**
